### DISCUSSION TITLE
CXSMILES Reaction parsing in Python

### DISCUSSION THREAD

**Discussion by JHucker** [rdkit/rdkit#5258] [Q&A] [answered]
From the PR https://github.com/rdkit/rdkit/pull/4895, RDkit appears to have functionality to parse CXSMILES reactions, however, I cannot find the python equivalent of RxnSmartsToChemicalReaction nor does ReactionFromSmarts seem to have CXSMILES parameters.

Does anyone know whether this is possible in Python and if so, how? Or is this purely a cpp feature currently? 

Motivation is that I want to be able to parse CXSMILES reactions and retain the fragment information, a (nonsense reaction) example: 

```
from rdkit.Chem import MolToSmiles
from rdkit.Chem.rdChemReactions import ReactionFromSmarts

rxn_cxsmiles = "[Na+].C1CCCCC1.[BH4-]>CCO>C1CCCCC1 |f:0.2|"
rxn = ReactionFromSmarts(rxn_cxsmiles, useSmiles=True)

print("\nReactants")
for mol in rxn.GetReactants():
    print(MolToSmiles(mol))      
```

so when this is parsed by rdkit.Chem.rdChemReactions.ReactionFromSmarts or equivalent, when looping through the reactants I would like to see ```[Na+].[BH4-]``` grouped. At present, the above excerpt will print the reactants in order separately i.e. ```[Na+]```, ```C1CCCCC1``` and then ```[BH4-]```.

**Comment by greglandrum** [comment]
Hi @JHucker: the `f` CXSMILES extension isn't currently supported by the RDKit.

If you would like to group reactants together in a reaction SMARTS, you can do so by including them in parentheses.
For your reaction this would be:
```
In [4]: rxn_cxsmiles = "([Na+].[BH4-]).C1CCCCC1>CCO>C1CCCCC1"
   ...: rxn = ReactionFromSmarts(rxn_cxsmiles, useSmiles=True)
   ...:
   ...: print("\nReactants")
   ...: for mol in rxn.GetReactants():
   ...:     print(MolToSmiles(mol))
   ...:

Reactants
[BH4-].[Na+]
C1CCCCC1
```

**Reply by JHucker** [reply]
Thanks for the information, Greg. I've attempted to group fragments via parenthesis before parsing, however it doesn't work for many reactions (~2/3rds from what I've sampled fail due to syntax errors). Here are some examples I've grouped which fail:

```
from rdkit.Chem.rdChemReactions import ReactionFromSmarts

grouped_list = [
  "[CH3:1][N:2]1[C:3]([CH2:4][CH2:5][C:6]1=O)=O>([Na+].[AlH4-]).Cl[Li]>[CH3:1][N:2]1[CH2:3][CH2:4][CH2:5][CH2:6]1",
  "[CH3:1][N:2]1[C:3]([CH2:4][CH2:5][C:6]1=O)=O>([Na+].[AlH4-])>[CH3:1][N:2]1[CH2:3][CH2:4][CH2:5][CH2:6]1",
  "[CH3:1][N:2]1[C:3]([CH2:4][CH2:5][C:6]1=O)=O>C1CCOC1.([Li+].[AlH4-])>[CH3:1][N:2]1[CH2:3][CH2:4][CH2:5][CH2:6]1",
  "CCOC(=O)C1C(N(C(CC1c1ccc(cc1)F)=O)C)=O>([Li+].[AlH4-])>CN1CCC(C(C1)CO)c1ccc(cc1)F",
  "Nc1ccccc1NC(=O)C1CCCN(C1)C(=O)OCc1ccccc1>O=C(Cl)Cl.(C1CCOC1.B)>O=C(OCc1ccccc1)N1CCCC(C1)Cn1c2ccccc2[nH]c1=O",
  "([OH2:1].[CH3:2][C:3]#[N:4])>([NH4+].CC([O-])=O)>([OH2:1].[CH3:2][C:3]#[N:4].O=C(O)C(F)(F)F)",
  "Cc1nc(c(c(n1)SC)C#N)O>(Cl.NO).c1ccncc1>Cc1nc(cc(n1)NO)O",
  "COC(=O)C1CCCCN(C1=O)Cc1ccccc1>C1CCOC1.([Li+].[AlH4-])>OCC1CCCCN(C1)Cc1ccccc1",
  "COC(=O)c1ccccc1I>CCN(CC)CC.CC#N.(CC([O-])=O.CC([O-])=O.[Pd+2])>COC(=O)c1ccccc1P(c1ccccc1)c1ccccc1",
  "Cn1cccc1C(=O)c1cc(ccc1Cl)[N+]([O-])=O>([NH4+].[OH-])>Cn1cccc1-c1nsc2ccc(cc12)[N+]([O-])=O"
]

parsed = []
for grouped_rxn in grouped_list:
    try:
        rxn = ReactionFromSmarts(grouped_rxn, useSmiles=True)
        assert rxn is not None
        parsed.append(rxn)
    except:
        continue

print(parsed)
```

These reactions can be successfully parsed if not grouped e.g. the equivalent for the first ```"[CH3:1][N:2]1[C:3]([CH2:4][CH2:5][C:6]1=O)=O>[Na+].[AlH4-].Cl[Li]>[CH3:1][N:2]1[CH2:3][CH2:4][CH2:5][CH2:6]1 |f:1.2|"```. 

I also note that components from these fail on Chem.MolFromSmiles too e.g. ```"(Cl.NO)"``` cannot be parsed, while ```"Cl.NO"``` can. 

Is there any issue with how these have been grouped / any alternative method available to retain fragment group information at a reaction level? 

FYI I'm using RDKit version 2022.03.2.

**Comment by greglandrum** [comment, marked as answer]
The grouping is not currently applied to the agents, so it will only work for reactants and products... this is likely why you are getting the syntax errors. 
The intention of the grouping in reactions is to allow you to say that the grouped fragments should all match in one reactant. This allows intramolecular reactions to be encoded in a straightforward way which otherwise wouldn't be possible.

Grouping using parentheses isn't supported in normal SMILES.

Before I try and suggest alternate solutions (assuming that there are any), what exactly are you trying to do?

**Reply by JHucker** [reply]
Good to know, thanks for that information. This is for reaction standardisation, where some optional steps are de-salting/ de-solventing and neutralising of defined fragments (per NextMove pistachio's dataset). This is being completed by chembl_structure_pipeline's standardizer.get_fragment_parent_mol. 

As RDKit cannot parse CXSMILES I'm currently using chython to read in the smiles and convert to RDF which RDKit parses and retains the fragments. While this is working 'okay', I'd like to eventually move to the one toolkit.

**Reply by greglandrum** [reply]
I have to confirm this, but I think it should be possible to add grouping support, i.e. using parentheses, to the parsing of the agents from the reaction smiles. Of course this would not show up until the 2022.09.1 RDKit release.

**Reply by JHucker** [reply]
For now grouping all components via parenthesis excluding the reagents will be sufficient, it would be excellent if there could be support for this in the future. Thank you for your support, Greg!
